tests/BaseCounterTest.php

```diff
@@ -0,0 +1,122 @@
+use PHPUnit\Framework\TestCase;
+use Yiisoft\Yii\RateLimiter\Counter;
+use Yiisoft\Yii\RateLimiter\Storage\StorageInterface;
+use Yiisoft\Yii\RateLimiter\Tests\Fixtures\FrozenTimeTimer;
+use Yiisoft\Yii\RateLimiter\Time\MicrotimeTimer;
+use Yiisoft\Yii\RateLimiter\Tests\Support\Assert;
+
+abstract class BaseCounterTest extends TestCase
+{
+    protected StorageInterface $storage;
+
+    abstract protected function createStorage(): StorageInterface;
+
+    protected function getStorage(): StorageInterface
+    {
+        if (empty($this->storage)) {
+            $this->storage = $this->createStorage();
+        }
+
+        return $this->storage;
+    }
+
+    public function testStatisticsShouldBeCorrectWhenLimitIsNotReached(): void
+    {
+        $counter = new Counter($this->getStorage(), 2, 5);
+
+        $statistics = $counter->hit('key');
+        $this->assertEquals(2, $statistics->getLimit());
+        $this->assertEquals(1, $statistics->getRemaining());
+        $this->assertGreaterThanOrEqual(time(), $statistics->getResetTime());
+        $this->assertFalse($statistics->isLimitReached());
+    }
+
+    public function testStatisticsShouldBeCorrectWhenLimitIsReached(): void
+    {
+        $counter = new Counter($this->getStorage(), 2, 4);
+
+        $statistics = $counter->hit('key');
+        $this->assertEquals(2, $statistics->getLimit());
+        $this->assertEquals(1, $statistics->getRemaining());
+        $this->assertGreaterThanOrEqual(time(), $statistics->getResetTime());
+        $this->assertFalse($statistics->isLimitReached());
+
+        $statistics = $counter->hit('key');
+        $this->assertEquals(2, $statistics->getLimit());
+        $this->assertEquals(0, $statistics->getRemaining());
+        $this->assertGreaterThanOrEqual(time(), $statistics->getResetTime());
+        $this->assertTrue($statistics->isLimitReached());
+    }
+
+    public function testShouldNotBeAbleToSetInvalidLimit(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        new Counter($this->getStorage(), 0, 60);
+    }
+
+    public function testShouldNotBeAbleToSetInvalidPeriod(): void
+    {
+        $this->expectException(InvalidArgumentException::class);
+        new Counter($this->getStorage(), 10, 0);
+    }
+
+    public function testIncrementMustBeUniformAfterLimitIsReached(): void
+    {
+        $timer = new FrozenTimeTimer();
+        $counter = new Counter(
+            $this->getStorage(),
+            10,
+            1,
+            86400,
+            'rate-limiter-',
+            $timer
+        );
+
+        // Start with the limit reached.
+        for ($i = 0; $i < 10; $i++) {
+            $counter->hit('key');
+        }
+
+        for ($i = 0; $i < 5; $i++) {
+            // Move timer forward for (period in milliseconds / limit)
+            // i.e. once in period / limit remaining allowance should be increased by 1.
+            FrozenTimeTimer::setTimeMark($timer->nowInMilliseconds() + 100);
```

Comment:
period / limit = 1s / 10 = 0.1s = 100ms.
For every 100ms increase, there is an additional chance of hitting.
> How about `+1` to be sure in precision?

Does it mean +1ms instead of +100ms?

Reply:
I think @xepozz meant `+ 101`. @xepozz am I correct?

Reply:
Got it, thanks